src/OpticalFlowEstimator.cpp

```diff
@@ -0,0 +1,609 @@
+    // we're moving in the -x direction
+    //
+    // Positive drdt means that the y vector rotates down, which means we think
+    // we're moving in the -y direction
+    //
+    // Note again that cos(roll) is negative, because the z axis is down
+    cv::Point2f correction_vel;
+    correction_vel.x = distance_to_plane
+                     * -dpitch_dt
+                     / std::cos(pitch);
+    correction_vel.y = distance_to_plane
+                     * -droll_dt
+                     / -std::cos(roll);
+
+    // Actual velocity in level camera frame (i.e. camera frame if our pitch
+    // and roll were zero)
+    Eigen::Vector3d corrected_vel(
+        velocity_uncorrected.x - correction_vel.x,
+        velocity_uncorrected.y - correction_vel.y,
+        0.0);
+
+    last_pitch = pitch;
+    last_roll = roll;
+
+    // Calculate covariance
+    Eigen::Matrix3d covariance = Eigen::Matrix3d::Zero();
+    covariance(0, 0) = std::pow(flow_estimator_settings_.variance_scale
+                              * dpitch_dt, 2.0)
+                      + flow_estimator_settings_.variance;
+    covariance(1, 1) = std::pow(flow_estimator_settings_.variance_scale
+                              * droll_dt, 2.0)
+                      + flow_estimator_settings_.variance;
+
+    // Rotation matrix from level camera frame to level_quad
+    Eigen::Matrix3d rotation_matrix;
+    rotation_matrix = Eigen::AngleAxisd(M_PI, Eigen::Vector3d::UnitX())
+                    * Eigen::AngleAxisd(-yaw, Eigen::Vector3d::UnitZ());
+
+    // Get velocity and covariance in level_quad frame
+    Eigen::Vector3d level_quad_vel = rotation_matrix * corrected_vel;
+    Eigen::Matrix3d level_quad_covariance = rotation_matrix
+                                          * covariance
+                                          * rotation_matrix.inverse();
+
+    // Correct for movement of camera relative to level_quad
+    geometry_msgs::PointStamped curr_pos, last_pos;
+    tf2::doTransform(curr_pos, curr_pos, current_camera_to_level_quad_tf_);
+    tf2::doTransform(last_pos, last_pos, last_camera_to_level_quad_tf_);
+    double camera_relative_vel_x = (curr_pos.point.x - last_pos.point.x) / dt;
+    double camera_relative_vel_y = (curr_pos.point.y - last_pos.point.y) / dt;
+    level_quad_vel.x() -= camera_relative_vel_x;
+    level_quad_vel.y() -= camera_relative_vel_y;
+
+    // Fill out the twist
+    geometry_msgs::TwistWithCovarianceStamped twist;
+    twist.header.stamp = time;
+    twist.header.frame_id = "level_quad";
+    twist.twist.twist.linear.x = level_quad_vel.x();
+    twist.twist.twist.linear.y = level_quad_vel.y();
+    twist.twist.twist.linear.z = 0.0;
+
+    twist.twist.covariance[0] = level_quad_covariance(0, 0);
+    twist.twist.covariance[1] = level_quad_covariance(0, 1);
+    twist.twist.covariance[6] = level_quad_covariance(1, 0);
+    twist.twist.covariance[7] = level_quad_covariance(1, 1);
+
+    // Publish intermediate twists for debugging
```

Comment:
In the spirit of consistency we should be publishing the camera_relative_vel_x and y and level_quad_vel too.

Reply:
Added publishers for camera_relative_vel and corrected_vel, level_quad_vel is pretty much what gets published so I didn't add that one

Reply:
Can we publish level_quad_vel? It could be important for debugging when rotating and we don't want to have to add it in later.